sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/Counter.java

```diff
@@ -295,6 +296,56 @@ public AggregationKind getKind() {
   public abstract CounterMean<T> getMean();
 
   /**
+   * CommitState represents whether counters' values have been committed to the backend.
+   *
+   * <p>Runners can use this information to optimize counters updates.
+   * For example, if counters are committed, runners may choose to skip the updates.
+   *
+   * <p>Counters' state transition table:
+   * Action\Current State         COMMITTED        DIRTY        COMMITTING
+   * add()                        DIRTY            DIRTY        DIRTY
+   * committing()                 None             COMMITTING   None
+   * committed()                  None             None         COMMITTED
+   */
+  @VisibleForTesting
+  enum CommitState {
+    COMMITTED,
+    DIRTY,
+    COMMITTING,
+  }
+
+  /**
+   * Returns if the counter contains non-committed aggregate.
+   */
+  public boolean isDirty() {
+    return commitState.get() != CommitState.COMMITTED;
+  }
+
+  /**
+   * Changes the counter from {@code CommitState.DIRTY} to {@code CommitState.COMMITTING}.
+   */
+  public boolean committing() {
+    return commitState.compareAndSet(CommitState.DIRTY, CommitState.COMMITTING);
+  }
+
+  /**
+   * Changes the counter from {@code CommitState.COMMITTING} to {@code CommitState.COMMITTED}.
+   */
+  public boolean committed() {
+    return commitState.compareAndSet(CommitState.COMMITTING, CommitState.COMMITTED);
+  }
+
+  /**
+   * Sets the counter to {@code CommitState.DIRTY}.
+   *
+   * <p>It needs to be called at the beginning of {@link #addValue}, {@link #resetToValue},
```

Comment:
"counter that is dirty but marked as committed" only happens inside addValue() and before finally block. And, eventually it will be correctly marked as dirty. There are also cases that counter that is committed but marked as dirty.
Both of them doesn't affect the eventual correctness.

added synchronized to get rid of the two cases above.